d2l-dom-focus.html

```diff
@@ -0,0 +1,160 @@
+
+			var composedChildren = D2L.Dom.getComposedChildren(node);
+			//
+
+			for (var i = 0; i < composedChildren.length; i++) {
+
+				if (composedChildren[i].nodeType !== 1) {
+					continue;
+				}
+
+				if (this.isFocusable(composedChildren[i], includeHidden)) {
+					return composedChildren[i];
+				}
+
+				var focusable = this.getFocusableDescendant(composedChildren[i], includeHidden);
+				if (focusable) {
+					return focusable;
+				}
+
+			}
+
+			return null;
+
+		},
+
+		getNextFocusable: function(node, includeHidden) {
+
+			if (!node) {
+				return null;
+			}
+
+			if (includeHidden === undefined) {
+				includeHidden = false;
+			}
+
+			var _getNextAnscestorSibling = function(node) {
+
+				var parentNode = D2L.Dom.getComposedParent(node);
+				while (parentNode) {
+
+					var nextParentSibling = parentNode.nextElementSibling;
+					if (nextParentSibling) {
+						return nextParentSibling;
+					}
+
+					parentNode = D2L.Dom.getComposedParent(parentNode);
+				}
+				return null;
+
+			};
+
+			var _getNextFocusable = function(node, ignore, ignoreChildren) {
+
+				if (!ignore && this.isFocusable(node, includeHidden)) {
+					return node;
+				}
+
+				if (!ignoreChildren) {
+					var focusable = this.getFocusableDescendant(node, includeHidden);
+					if (focusable) {
+						return focusable;
+					}
+				}
+
+				var nextSibling = node.nextElementSibling;
+				if (nextSibling) {
+					var siblingFocusable = _getNextFocusable(nextSibling, false, false);
+					if (siblingFocusable) {
+						return siblingFocusable;
+					}
+
+					return null;
+				}
+
+				var nextParentSibling = _getNextAnscestorSibling(node);
+				if (nextParentSibling) {
+					var parentSibingFocusable = _getNextFocusable(nextParentSibling, false, false);
+					if (parentSibingFocusable) {
+						return parentSibingFocusable;
+					}
+				}
+
+				return null;
+
+			}.bind(this);
+
+			var focusable = _getNextFocusable(node, true, false);
+			return focusable;
+
+		},
+
+		isFocusable: function(node, includeHidden) {
+
+			if (!node || node.disabled) {
+				return false;
+			}
+
+			var _isFocusable = false;
+
+			// IE treats all nodes without tabindex explicitly set as having tabindex=0
+			if (node.getAttributeNode) {
+				var tabIndexAttr = node.getAttributeNode('tabindex');
+				if (tabIndexAttr && tabIndexAttr.specified && tabIndexAttr.value >= 0) {
+					_isFocusable = true;
+				}
+			}
+
+			if (!_isFocusable) {
+				var nodeName = node.nodeName.toLowerCase();
+				if (this._focusableElements[nodeName]) {
+					_isFocusable = true;
+				}
+			}
+
+			if (_isFocusable && !includeHidden) {
+				// only perform visibility check if absolutely necessary
+				if (!D2L.Dom.Visibility.isVisible(node)) {
+					return false;
+				}
+			}
+
+			return _isFocusable;
+
+		},
+
+		getPreviousFocusable: function() {
+
+		}
+
+	};
+
+	window.D2L = window.D2L || {};
+	window.D2L.Dom = window.D2L.Dom || {};
+	window.D2L.Dom.Focus = Focus;
```

Comment:
Initially I had exposed these functions under the banner of _behaviors_.  These functions are all stateless, and they aren't really behaviors of web components so I've kept them separate.  For such cases it is still recommended to `import` as HTML file so that it can be deduplicated (interesting discussion about that [here](http://stackoverflow.com/questions/22135095/polymer-element-with-javascript-dependencies)).  I wasn't sure what name to give them... simply started with this.